### PR TITLE
Various performance improvements.

### PR DESCRIPTION
**Pull request by millems**
Benchmarks in progress.

### CODE REVIEW

File: http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java [hunk 1]
```diff
@@ -172,12 +169,35 @@ public String toString() {
...
-        private Map<String, List<String>> headers = new LinkedHashMap<>();
+
+        private boolean headersAreFromToBuilder;
+        private Map<String, List<String>> headers;
+
         private ContentStreamProvider contentStreamProvider;
 
         Builder() {
+            queryParameters = new LinkedHashMap<>();
+            queryParametersAreFromToBuilder = false;
```
Comment (cenedhryn) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java:185]: Substance over beauty!

Comment (millems) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java:185]: Hah, I wrote a class to do this originally, but it just made things harder to read. If we find ourselves doing this a lot we can put it somewhere else...

File: core/auth/src/main/java/software/amazon/awssdk/auth/signer/internal/AbstractAws4Signer.java [hunk 2]
```diff
@@ -191,26 +204,42 @@ protected abstract void processRequestPayload(SdkHttpFullRequest.Builder mutable
...
-                                  getSignedHeadersString(request.headers()) +
+                                  signedHeadersString +
                                   SignerConstant.LINE_SEPARATOR +
                                   contentSha256;
 
         LOG.trace(() -> "AWS4 Canonical Request: " + canonicalRequest);
         return canonicalRequest;
     }
 
+    private Map<String, List<String>> canonicalizeSigningHeaders(Map<String, List<String>> headers) {
```
[the remark this answers is not in the document]
Comment (millems) [reply core/auth/src/main/java/software/amazon/awssdk/auth/signer/internal/AbstractAws4Signer.java]: Fixed!

**Comment by millems** [comment]
Before Change:
```
Benchmark                                                                               (sslProviderValue)  (testItem)    Mode     Cnt        Score        Error  Units
s.a.a.b.apicall.httpclient.async.NettyClientH1NonTlsBenchmark.concurrentApiCall                        N/A         N/A   thrpt      10    11245.811 ?    287.866  ops/s
s.a.a.b.apicall.httpclient.async.NettyClientH1NonTlsBenchmark.sequentialApiCall                        N/A         N/A   thrpt      10     3197.988 ?     30.884  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH1Benchmark.concurrentApiCall                          jdk         N/A   thrpt      10    10368.854 ?    152.575  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH1Benchmark.concurrentApiCall                      openssl         N/A   thrpt      10    10605.050 ?    107.683  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH1Benchmark.sequentialApiCall                          jdk         N/A   thrpt      10     2756.870 ?     27.622  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH1Benchmark.sequentialApiCall                      openssl         N/A   thrpt      10     2821.305 ?     60.341  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH2Benchmark.concurrentApiCall                          jdk         N/A   thrpt      10     6242.370 ?     67.031  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH2Benchmark.concurrentApiCall                      openssl         N/A   thrpt      10     7132.845 ?     72.225  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH2Benchmark.sequentialApiCall                          jdk         N/A   thrpt      10     2285.479 ?     30.133  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH2Benchmark.sequentialApiCall                      openssl         N/A   thrpt      10     2396.177 ?     17.290  ops/s
s.a.a.b.apicall.httpclient.sync.ApacheHttpClientBenchmark.concurrentApiCall                            N/A         N/A   thrpt      10    15902.702 ?     93.379  ops/s
s.a.a.b.apicall.httpclient.sync.ApacheHttpClientBenchmark.sequentialApiCall                            N/A         N/A   thrpt      10     3880.689 ?     96.634  ops/s
s.a.a.b.apicall.httpclient.sync.UrlConnectionHttpClientBenchmark.sequentialApiCall                     N/A         N/A   thrpt      10      765.424 ?     15.950  ops/s
s.a.a.b.apicall.protocol.Ec2ProtocolBenchmark.successfulResponse                                       N/A         N/A   thrpt      10     9510.425 ?     35.961  ops/s
s.a.a.b.apicall.protocol.JsonProtocolBenchmark.successfulResponse                                      N/A         N/A   thrpt      10    15343.291 ?    106.156  ops/s
s.a.a.b.apicall.protocol.QueryProtocolBenchmark.successfulResponse                                     N/A         N/A   thrpt      10    10340.202 ?    100.980  ops/s
s.a.a.b.apicall.protocol.XmlProtocolBenchmark.successfulResponse                                       N/A         N/A   thrpt      10     8670.604 ?    109.115  ops/s
```

After Change:
```
Benchmark                                                                               (sslProviderValue)  (testItem)    Mode     Cnt        Score        Error  Units
s.a.a.b.apicall.httpclient.async.NettyClientH1NonTlsBenchmark.concurrentApiCall                        N/A         N/A   thrpt      10    14122.509 ?    131.076  ops/s
s.a.a.b.apicall.httpclient.async.NettyClientH1NonTlsBenchmark.sequentialApiCall                        N/A         N/A   thrpt      10     3457.482 ?     30.247  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH1Benchmark.concurrentApiCall                          jdk         N/A   thrpt      10    12422.813 ?    138.608  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH1Benchmark.concurrentApiCall                      openssl         N/A   thrpt      10    12613.108 ?    318.684  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH1Benchmark.sequentialApiCall                          jdk         N/A   thrpt      10     2975.029 ?     18.522  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH1Benchmark.sequentialApiCall                      openssl         N/A   thrpt      10     3043.285 ?     56.311  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH2Benchmark.concurrentApiCall                          jdk         N/A   thrpt      10     6585.652 ?     70.420  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH2Benchmark.concurrentApiCall                      openssl         N/A   thrpt      10     7561.217 ?     97.082  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH2Benchmark.sequentialApiCall                          jdk         N/A   thrpt      10     2507.358 ?     16.469  ops/s
s.a.a.b.apicall.httpclient.async.NettyHttpClientH2Benchmark.sequentialApiCall                      openssl         N/A   thrpt      10     2613.694 ?     17.699  ops/s
s.a.a.b.apicall.httpclient.sync.ApacheHttpClientBenchmark.concurrentApiCall                            N/A         N/A   thrpt      10    17899.993 ?    217.164  ops/s
s.a.a.b.apicall.httpclient.sync.ApacheHttpClientBenchmark.sequentialApiCall                            N/A         N/A   thrpt      10     4346.483 ?    103.673  ops/s
s.a.a.b.apicall.httpclient.sync.UrlConnectionHttpClientBenchmark.sequentialApiCall                     N/A         N/A   thrpt      10      798.077 ?      5.132  ops/s
s.a.a.b.apicall.protocol.Ec2ProtocolBenchmark.successfulResponse                                       N/A         N/A   thrpt      10    11387.648 ?    191.272  ops/s
s.a.a.b.apicall.protocol.JsonProtocolBenchmark.successfulResponse                                      N/A         N/A   thrpt      10    19420.380 ?    324.009  ops/s
s.a.a.b.apicall.protocol.QueryProtocolBenchmark.successfulResponse                                     N/A         N/A   thrpt      10    12269.337 ?    417.780  ops/s
s.a.a.b.apicall.protocol.XmlProtocolBenchmark.successfulResponse                                       N/A         N/A   thrpt      10    10276.002 ?    138.007  ops/s
```

**Pull request merged**